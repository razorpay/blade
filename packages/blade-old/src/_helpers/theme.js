import { spacings, colors } from '../tokens';
import theme from '../tokens/theme';

const makePxValue = (value) => {
  const values = [].concat(value);
  if (values.length > 4) {
    throw new Error('Error in makePxValue: array length should be less than or equal to 4');
  }
  return values.map((v) => (typeof v === 'string' ? v : `${v * spacings.unit}px`)).join(' ');
};

const getColorKeys = () => {
  return Object.keys(theme.bladeOld.colors)
    .map((base) => {
      return Object.keys(theme.bladeOld.colors[base]).map((shade) => `${base}.${shade}`);
    })
    .reduce((acc, val) => [...acc, ...val]);
};

const getColor = (currentTheme, color) => {
  const [base, shade] = color.split('.');
  try {
    return currentTheme.bladeOld.colors[base][shade];
  } catch {
    throw new Error(
      `Invalid color prop\n\nReceived color: ${color}\n\nValid color values:\n ${getColorKeys()}`,
    );
  }
};

const getVariantColorKeys = () => Object.keys(theme.bladeOld.colors);

const getLineHeight = (currentTheme, textSize) => {
  switch (textSize) {
    case 'xxsmall':
      return currentTheme.bladeOld.fonts.lineHeight.xsmall;
    case 'xsmall':
      return currentTheme.bladeOld.fonts.lineHeight.small;
    case 'small':
      return currentTheme.bladeOld.fonts.lineHeight.small;
    case 'medium':
      return currentTheme.bladeOld.fonts.lineHeight.medium;
    case 'xxxlarge':
      return currentTheme.bladeOld.fonts.lineHeight.xxlarge;
    case 'xxlarge':
      return currentTheme.bladeOld.fonts.lineHeight.xxlarge;
    case 'xlarge':
      return currentTheme.bladeOld.fonts.lineHeight.xlarge;
    case 'large':
    default:
      return currentTheme.bladeOld.fonts.lineHeight.large;
  }
};

const getThemeColors = () => {
  const getColors = (obj) => {
    return Object.keys(obj).reduce((themeColors, key) => {
      if (typeof obj[key] === 'object') {
        return themeColors.concat(getColors(obj[key]));
      } else themeColors.push(obj[key]);
      return themeColors;
    }, []);
  };
  return getColors(colors);
};

export { makePxValue, getColorKeys, getColor, getVariantColorKeys, getLineHeight, getThemeColors };
