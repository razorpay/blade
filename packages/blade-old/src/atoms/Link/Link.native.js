import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity } from 'react-native';
import Text from '../Text';
import baseTheme from '../../tokens/theme';
import automation from '../../_helpers/automation-attributes';
import Flex from '../Flex';

const Link = ({ children, onClick, size, weight, hitSlop, testID }) => {
  const [active, setActive] = useState(false);
  const inactiveTextColor = 'primary.800';
  const activeTextColor = 'primary.600';

  const onPressIn = useCallback(() => {
    setActive(true);
  }, [setActive]);
  const onPressOut = useCallback(() => {
    setActive(false);
  }, [setActive]);

  return (
    <Flex alignSelf="flex-start">
      <TouchableOpacity
        activeOpacity={1}
        onPress={onClick}
        onPressIn={onPressIn}
        onPressOut={onPressOut}
        hitSlop={hitSlop}
        {...automation(testID)}
      >
        <Text
          color={active ? activeTextColor : inactiveTextColor}
          size={size}
          weight={weight}
          _isUnderlined={active}
        >
          {children}
        </Text>
      </TouchableOpacity>
    </Flex>
  );
};

Link.propTypes = {
  children: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  size: PropTypes.oneOf(['xxsmall', 'xsmall', 'small', 'medium', 'large']),
  weight: PropTypes.oneOf(Object.keys(baseTheme.fonts.weight)),
  testID: PropTypes.string,
  hitSlop: PropTypes.shape({
    left: PropTypes.number,
    top: PropTypes.number,
    right: PropTypes.number,
    bottom: PropTypes.number,
  }),
};

Link.defaultProps = {
  onClick: () => {},
  size: 'medium',
  weight: 'regular',
  testID: 'ds-link',
};

export default Link;
