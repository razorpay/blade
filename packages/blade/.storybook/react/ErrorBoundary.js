import React from 'react';

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, errorMessage: '' };
  }

  static getDerivedStateFromError(error) {
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    this.setState({ errorMessage: error.message });
  }

  render() {
    if (this.state.hasError) {
      return (
        <div>
          Please refresh the page. Following error has occurred:
          <div
            style={{
              color: 'red',
            }}
          >
            {this.state.errorMessage}
          </div>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
