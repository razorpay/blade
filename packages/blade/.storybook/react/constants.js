export const INTERNAL_STORY_ADDON_PARAM = 'showInternalComponents';
