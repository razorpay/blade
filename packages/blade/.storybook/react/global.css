* {
  /* Need this to ensure mdx stories don't break visually */
  box-sizing: border-box;
}

.sbdocs-h1,
.sbdocs-h2,
.sbdocs-h3,
.sbdocs-h4,
.sbdocs-h5,
.sbdocs-h6 {
  font-family: 'TASA Orbiter', -apple-system, BlinkMacSystemFont, San Francisco, Segoe UI, Roboto,
    Helvetica Neue, sans-serif !important;
  font-weight: 600 !important;
}

body {
  width: 100%;
  height: 100%;
  padding: 0 !important;
}
