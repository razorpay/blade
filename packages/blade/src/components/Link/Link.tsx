import type { ReactElement } from 'react';
import styled from 'styled-components';
import type { DurationString, EasingString } from '../../tokens/global/motion';
import type { ActionStates } from '../../tokens/theme/theme.d';
import getIn from '../../utils/getIn';
import makeBorderSize from '../../utils/makeBorderSize';
import makeMotionTime from '../../utils/makeMotionTime';
import type { DotNotationSpacingStringToken } from '../../_helpers/types';
import type { Theme } from '../BladeProvider';
import { useTheme } from '../BladeProvider';
import Box from '../Box';
import type { IconComponent, IconProps } from '../Icons';

import type { BaseTextProps } from '../Typography/BaseText';
import BaseText from '../Typography/BaseText';
import StyledLink from './StyledLink.web';
import useInteraction from './useInteraction';

type LinkCommonProps = {
  variant?: 'anchor' | 'button';
  icon?: IconComponent;
  iconPosition?: 'left' | 'right';
  isDisabled?: boolean;
  onClick?: () => void;
  href?: string;
  target?: string;
};

type LinkWithoutIconProps = LinkCommonProps & {
  icon?: undefined;
  children: string;
};

/*
   Optional children prop when icon is provided
  */
type LinkWithIconProps = LinkCommonProps & {
  icon: IconComponent;
  children?: string;
};

export type LinkProps = LinkWithIconProps | LinkWithoutIconProps;

type LinkStyleProps = {
  as: 'a' | 'button';
  textDecoration: 'underline' | 'none';
  iconColor: IconProps['color'];
  iconPadding: DotNotationSpacingStringToken;
  textColor: BaseTextProps['color'];
  focusRingColor: string;
  motionDuration: DurationString;
  motionEasing: EasingString;
};

const getProps = ({
  theme,
  variant,
  currentInteraction,
}: {
  theme: Theme;
  variant: NonNullable<LinkCommonProps['variant']>;
  currentInteraction: keyof ActionStates;
}): LinkStyleProps => {
  const props: LinkStyleProps = {
    as: variant === 'anchor' ? 'a' : 'button',
    textDecoration: variant === 'anchor' && currentInteraction !== 'default' ? 'underline' : 'none',
    iconColor: `action.icon.link.${currentInteraction}`,
    iconPadding: 'spacing.1',
    textColor: `action.text.link.${currentInteraction}`,
    focusRingColor:
      currentInteraction == 'focus' || currentInteraction == 'active'
        ? getIn(theme.colors, 'brand.primary.400')
        : 'none',
    motionDuration: 'duration.2xquick',
    motionEasing: 'easing.standard.effective',
  };
  return props;
};

const StyledBox = styled(Box)<{
  focusRingColor: string;
  motionDuration: DurationString;
  motionEasing: EasingString;
}>(({ theme, focusRingColor, motionEasing, motionDuration }) => ({
  boxShadow: `0px 0px 0px 4px ${focusRingColor}`,
  borderRadius: makeBorderSize(theme.border.radius.small),
  transitionProperty: 'box-shadow',
  transitionTimingFunction: getIn(theme.motion, motionEasing),
  transitionDuration: makeMotionTime(getIn(theme.motion, motionDuration)),
  '*': {
    transitionProperty: 'color, fill',
    transitionTimingFunction: getIn(theme.motion, motionEasing),
    transitionDuration: makeMotionTime(getIn(theme.motion, motionDuration)),
  },
}));

const Link = ({
  children,
  icon: Icon,
  iconPosition = 'left',
  isDisabled = false,
  onClick,
  variant = 'anchor',
  href,
  target,
}: LinkProps): ReactElement => {
  console.log('unused props', isDisabled);
  const { currentInteraction, ...syntheticEvents } = useInteraction();
  const { theme } = useTheme();
  const {
    as,
    textDecoration,
    iconColor,
    iconPadding,
    textColor,
    focusRingColor,
    motionDuration,
    motionEasing,
  } = getProps({
    theme,
    variant,
    currentInteraction,
  });

  return (
    <StyledLink {...syntheticEvents} as={as} href={href} target={target} onClick={onClick}>
      <StyledBox
        display="flex"
        flexDirection="row"
        width="fit-content"
        justifyContent="center"
        alignItems="center"
        focusRingColor={focusRingColor}
        motionDuration={motionDuration}
        motionEasing={motionEasing}
      >
        {Icon && iconPosition == 'left' ? (
          <Box paddingRight={iconPadding}>
            <Icon color={iconColor} size="xsmall" />
          </Box>
        ) : null}
        <BaseText textDecorationLine={textDecoration} color={textColor} fontSize={100}>
          {children}
        </BaseText>
        {Icon && iconPosition == 'right' ? (
          <Box paddingLeft={iconPadding} display="flex" alignSelf="center">
            <Icon color={iconColor} size="xsmall" />
          </Box>
        ) : null}
      </StyledBox>
    </StyledLink>
  );
};

export default Link;
