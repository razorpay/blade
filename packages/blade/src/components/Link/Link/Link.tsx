import type { ReactElement, SyntheticEvent } from 'react';
import React from 'react';
import type { BaseLinkProps } from '../BaseLink';
import { BaseLink } from '../BaseLink';
import type { IconComponent } from '~components/Icons';
import type { StyledPropsBlade } from '~components/Box/styledProps';
import { getStyledProps } from '~components/Box/styledProps';
import type { StringChildrenType, TestID, BladeElementRef } from '~utils/types';
import type { Platform } from '~utils';
import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
import type { BladeCommonEvents } from '~components/types';

type LinkCommonProps = {
  variant?: 'anchor' | 'button';
  icon?: IconComponent;
  color?: 'default' | 'white' | 'neutral';
  iconPosition?: 'left' | 'right';
  isDisabled?: boolean;
  onClick?: (event: SyntheticEvent) => void;
  href?: string;
  target?: string;
  accessibilityLabel?: string;

  /**
   * It is exposed for internal usage with tooltip.
   *
   * @private
   */
  'aria-describedby'?: string;

  /**
   * Sets the size of the link
   *
   * @default medium
   */
  size?: BaseLinkProps['size'];
} & TestID &
  StyledPropsBlade &
  BladeCommonEvents &
  Platform.Select<{
    native: {
      /**
       * Defines how far your touch can start away from the link
       */
      hitSlop?:
        | {
            top?: number;
            right?: number;
            bottom?: number;
            left?: number;
          }
        | number;
      /**
       * This is a web only prop and has no effect on react-native.
       */
      htmlTitle?: undefined;
    };
    web: {
      /**
       * This is a react-native only prop and has no effect on web.
       */
      hitSlop?: undefined;
      /**
       * The title of the link which is displayed as a tooltip.
       */
      htmlTitle?: string;
    };
  }>;

/*
  Mandatory children prop when icon is not provided
*/
type LinkWithoutIconProps = LinkCommonProps & {
  icon?: undefined;
  children: StringChildrenType;
};

/*
  Optional children prop when icon is provided
*/
type LinkWithIconProps = LinkCommonProps & {
  icon: IconComponent;
  children?: StringChildrenType;
};

/*
  Link Props with or without an icon
*/
type LinkPropsWithOrWithoutIcon = LinkWithIconProps | LinkWithoutIconProps;

/*
  Link Props when variant is anchor
*/
type LinkAnchorVariantProps = LinkPropsWithOrWithoutIcon & {
  variant?: 'anchor';
  href?: string;
  target?: string;
  rel?: string;
  isDisabled?: undefined;
};

/*
  Link Props when variant is button
*/
export type LinkButtonVariantProps = LinkPropsWithOrWithoutIcon & {
  variant?: 'button';
  isDisabled?: boolean;
  href?: undefined;
  target?: undefined;
  rel?: undefined;
};

/*
  Link Props when variant is anchor or button
*/
export type LinkProps = LinkAnchorVariantProps | LinkButtonVariantProps;

const _Link: React.ForwardRefRenderFunction<BladeElementRef, LinkProps> = (
  {
    children,
    icon,
    iconPosition = 'left',
    isDisabled = false,
    onClick,
    variant = 'anchor',
    color = 'default',
    href,
    target,
    rel,
    accessibilityLabel,
    size = 'medium',
    testID,
    hitSlop,
    htmlTitle,
    onBlur,
    onFocus,
    onMouseLeave,
    onMouseMove,
    onPointerDown,
    onPointerEnter,
    onTouchStart,
    onTouchEnd,
    ...rest
  },
  ref,
): ReactElement => {
  return (
    <BaseLink
      {...(icon ? { icon, children } : { children })}
      {...(variant === 'anchor' ? { variant, href, target, rel } : { variant, isDisabled })}
      ref={ref as never}
      iconPosition={iconPosition}
      onClick={onClick}
      accessibilityProps={{ label: accessibilityLabel, describedBy: rest['aria-describedby'] }}
      size={size}
      color={color}
      testID={testID}
      hitSlop={hitSlop}
      htmlTitle={htmlTitle}
      onBlur={onBlur}
      onFocus={onFocus}
      onMouseLeave={onMouseLeave}
      onMouseMove={onMouseMove}
      onPointerDown={onPointerDown}
      onPointerEnter={onPointerEnter}
      onTouchStart={onTouchStart}
      onTouchEnd={onTouchEnd}
      {...getStyledProps(rest)}
    />
  );
};

const Link = assignWithoutSideEffects(React.forwardRef(_Link), {
  displayName: 'Link',
  componentId: 'Link',
});

export default Link;
