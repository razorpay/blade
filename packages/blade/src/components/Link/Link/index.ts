export { default as Link } from './Link';
export * from './Link';
