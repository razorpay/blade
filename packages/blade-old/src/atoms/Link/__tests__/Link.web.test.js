import React from 'react';
import { fireEvent } from '@testing-library/react';
import Link from '../index';
import { renderWithTheme } from '../../../_helpers/testing';

describe('<Link />', () => {
  describe('Size', () => {
    it('should render a small Link', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(<Link size="small">{displayText}</Link>);
      expect(container).toMatchSnapshot();
    });

    it('should render a medium Link', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(<Link size="medium">{displayText}</Link>);
      expect(container).toMatchSnapshot();
    });

    it('should render a large Link', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(<Link size="large">{displayText}</Link>);
      expect(container).toMatchSnapshot();
    });
  });
  describe('Selectors', () => {
    it('should render a Link with hover styles', () => {
      const displayText = 'Displaying some link';
      const mockOnPress = jest.fn();
      const { getByText } = renderWithTheme(<Link>{displayText}</Link>);
      const link = getByText(displayText).parentNode;
      fireEvent.mouseOver(link);
      expect(mockOnPress).toMatchSnapshot();
    });
    it('should render a Link with focus styles', () => {
      const displayText = 'Displaying some link';
      const mockOnPress = jest.fn();
      const { getByText } = renderWithTheme(<Link>{displayText}</Link>);
      const link = getByText(displayText).parentNode;
      fireEvent.focus(link);
      expect(mockOnPress).toMatchSnapshot();
    });
    it('should render a Link with active styles', () => {
      const displayText = 'Displaying some link';
      const mockOnPress = jest.fn();
      const { getByText } = renderWithTheme(<Link>{displayText}</Link>);
      const link = getByText(displayText).parentNode;
      fireEvent.click(link);
      expect(mockOnPress).toMatchSnapshot();
    });
  });
  describe('various Link props', () => {
    it('should render a Link with default props', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(<Link>{displayText}</Link>);
      expect(container).toMatchSnapshot();
    });
    it('should a render a Link with disabled prop', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(<Link disabled>{displayText}</Link>);
      expect(container).toMatchSnapshot();
    });
    it('should render a Link with href attr', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(
        <Link size="large" href="https://www.razorpay.com">
          {displayText}
        </Link>,
      );
      expect(container).toMatchSnapshot();
    });
    it('should reender a Link with target attr', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(
        <Link
          size="large"
          href="https://www.razorpay.com"
          target="_blank"
          rel="noreferrer noopener"
        >
          {displayText}
        </Link>,
      );
      expect(container).toMatchSnapshot();
    });
    it('should render a Link with rel attribute', () => {
      const displayText = 'Displaying some link';
      const { container } = renderWithTheme(
        <Link
          size="large"
          href="https://www.razorpay.com"
          target="_blank"
          rel="noreferrer noopener"
        >
          {displayText}
        </Link>,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
