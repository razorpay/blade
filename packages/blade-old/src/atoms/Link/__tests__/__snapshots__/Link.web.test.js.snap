// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> Selectors should render a Link with active styles 1`] = `[MockFunction]`;

exports[`<Link /> Selectors should render a Link with focus styles 1`] = `[MockFunction]`;

exports[`<Link /> Selectors should render a Link with hover styles 1`] = `[MockFunction]`;

exports[`<Link /> Size should render a large Link 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 gQrLKz"
    color="shade.980"
    data-testid="ds-text"
    target="_self"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> Size should render a medium Link 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 biSdwx"
    color="shade.980"
    data-testid="ds-text"
    target="_self"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> Size should render a small Link 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 bcybCt"
    color="shade.980"
    data-testid="ds-text"
    target="_self"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> various Link props should a render a Link with disabled prop 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 jCGuPd"
    color="shade.980"
    data-testid="ds-text"
    disabled=""
    href=""
    target="_self"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> various Link props should reender a Link with target attr 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 gQrLKz"
    color="shade.980"
    data-testid="ds-text"
    href="https://www.razorpay.com"
    target="_blank"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> various Link props should render a Link with default props 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 biSdwx"
    color="shade.980"
    data-testid="ds-text"
    target="_self"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> various Link props should render a Link with href attr 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 gQrLKz"
    color="shade.980"
    data-testid="ds-text"
    href="https://www.razorpay.com"
    target="_self"
  >
    Displaying some link
  </a>
</div>
`;

exports[`<Link /> various Link props should render a Link with rel attribute 1`] = `
<div>
  <a
    class="Textweb__Text-sc-1qrdpx3-0 Linkweb__StyledText-c2kinr-0 gQrLKz"
    color="shade.980"
    data-testid="ds-text"
    href="https://www.razorpay.com"
    rel="noreferrer noopener"
    target="_blank"
  >
    Displaying some link
  </a>
</div>
`;
